File: 2022/Day6.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import java.util.ArrayList;

public class Day6 {
    public static void main(String[] args) throws FileNotFoundException{
        Day6 solution = new Day6();
        System.out.println("Part 1 Answer: " + solution.part1(new Scanner(new File("input.txt"))));
        System.out.println("Part 2 Answer: " + solution.part2(new Scanner(new File("input.txt"))));
    }

    public int part1(Scanner input) {
        String data = input.nextLine();
        for (int i = 0; i < data.length()-3; i++) {
            ArrayList<Character> marker = new ArrayList<Character>();
            for (int j = 0; j < 4; j++) {
                if (marker.contains(data.charAt(i + j)))
                    break;
                marker.add(data.charAt(i + j));
            }
            if (marker.size() == 4)
                return i + 4;
        }
        return -1;
    }

    public int part2(Scanner input) {
        String data = input.nextLine();
        for (int i = 0; i < data.length()-13; i++) {
            ArrayList<Character> marker = new ArrayList<Character>();
            for (int j = 0; j < 14; j++) {
                if (marker.contains(data.charAt(i + j)))
                    break;
                marker.add(data.charAt(i + j));
            }
            if (marker.size() == 14)
                return i + 14;
        }
        return -1;
    }
}

File: 2021/Day6.py
def main():
    with open("input.txt") as f:
        input = f.read()
        print(f"Part 1 Answer: {part1(input)}")
        print(f"Part 2 Answer: {part2(input)}")

def part1(input):
    fish = [int(x) for x in input.split(",")]
    for i in range(80):
        for j in range(len(fish)):
            if fish[j] == 0:
                fish[j] = 6
                fish.append(8)
            else:
                fish[j] -= 1
    return len(fish)

def part2(input):
    fishGroups = [0] * 9
    for fish in input.split(","):
        fishGroups[int(fish)] += 1
    for i in range(256):
        new = fishGroups[0]
        for j in range(len(fishGroups) - 1):
            fishGroups[j] = fishGroups[j + 1]
        fishGroups[len(fishGroups) - 1] = new
        fishGroups[6] += new
    return sum(fishGroups)

if __name__ == '__main__':
    main()